intermediate_source/ddp_minGPT.rst

```diff
@@ -0,0 +1,93 @@
+
+We use `hydra <https://hydra.cc/>` to centrally manage all the configurations for our training run.  
+Once the code has been refactored, we run it first on a single-node with 4 GPUs, and then on a slurm cluster.
+
+
+.. grid:: 2
+
+   .. grid-item-card:: :octicon:`mortar-board;1em;` What you will learn
+      :shadow: none
+
+      -  Refactor a (nicely structured) project to use DDP training
+      -  Best practices when writing a distributed training script
+      -  Increased flexibility with saving/loading artifacts in the cloud
+      -  When is DDP NOT suitable
+
+
+   .. grid-item-card:: :octicon:list-unordered;1em;` Prerequisites
+      :shadow: none
+
+      * Familiarity with `multi-GPU training <beginner/ddp_multigpu.html>`__ and `torchrun <beginner/ddp_fault_tolerance.html>`__ 
+      * [Optional] Familiarity with `multinode training <ddp_multinode.html>`__
+      * 2 or more TCP-reachable GPU machines (this tutorial uses AWS p3.2xlarge instances)
+      * PyTorch `installed <https://pytorch.org/get-started/locally/>`__ with CUDA on all machines
+
+
+
+
+
+View the code used in this video: https://github.com/suraj813/minGPT-ddp
+
+
+.. raw:: html
+
+   <div style="margin-top:10px; margin-bottom:10px;">
+     <iframe width="560" height="315" src="https://www.youtube.com/embed/XFsFDGKZHh4" frameborder="0" allow="accelerometer; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+   </div>
+
+
+
+Files used for training
+~~~~~~~~~~~~~~~~~~~~~~~~
+- `trainer.py <https://github.com/pytorch/examples/tree/main/distributed/minGPT-ddp/mingpt/trainer.py>` includes the Trainer class that runs the distributed training iterations on the model
+with the provided dataset.
+- `model.py <https://github.com/pytorch/examples/tree/main/distributed/minGPT-ddp/mingpt/model.py>` defines the model architecture.
+- `char_dataset.py <https://github.com/pytorch/examples/tree/main/distributed/minGPT-ddp/mingpt/char_dataset.py>` contains the `Dataset`class for a character-level dataset.
+- `gpt2_train_cfg.yaml <https://github.com/pytorch/examples/tree/main/distributed/minGPT-ddp/mingpt/gpt2_train_cfg.yaml>` contains the configurations for data, model, optimizer and training run.
+- `main.py <https://github.com/pytorch/examples/tree/main/distributed/minGPT-ddp/mingpt/main.py>` is the entry point to the trainig job. 
+It sets up the DDP process group, reads all the configurations and runs the training job.
+
+
+Saving and Loading from the cloud
+~~~~~~~~~~~~~~~~~~~~~~~~
+In the video, we save training snapshots directly to the cloud. This gives us the flexibility to continue training
+from any node that has access to the cloud bucket.
+
+
+Using Mixed Precision
+~~~~~~~~~~~~~~~~~~~~~~~~
+To speed things up, you might be able to use `Mixed Precision <https://pytorch.org/docs/stable/amp.html>`__ to train your models. 
+In Mixed Precision, some parts of the training process are carried out in FP16 half-precision, while other steps 
```

Comment:
"carried out in reduced precision" as amp now supports other types than just fp17